decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,476 @@
+
+The library must be designed to simplify integration into current
+_and_ future programming environments, without the need to re-implement
+critical business logic.
+
+## Relevant data
+
+### Required language support
+
+`C++`, `Python`
+
+### Projected language support
+
+`C`, `JavaScript`, `Java`, `Swift`, `Rust`, `go`
+
+### Bindings
+
+Research has shown that the common denominator when addressing the above
+listed languages is a `C` interface. A standard `C` API surface, with
+clear and manageable ownership semantics would facilitate support in the
+broadest range of future programming environments.
+
+Ideally, any bindings would also allow idiomatic programming within
+the target language, rather than forcing specific design patterns from
+any internals of OpenAssetIO - such as synchronous programming, RAII,
+etc.
+
+### Performance
+
+A core requirement raised by numerous parties is that the "happy path"
+should be as performant as possible. Thus, it is required that any
+mechanism introduced to broaden compatibility doesn't compromise the
+(current) majority deployment scenarios.
+
+Of note, a `C++` host talking to a `C++` manager has been the most
+frequently requested high-performance pathway.
+
+### Development lifecycle
+
+The overhead of maintaining dynamically linked `C++` libraries within a
+plugin-based architecture is often fraught with constantly chasing ABI
+changes and/or conflicts between mutual dependencies.
+
+Foundry has prior art in using `C` function pointer suites as a compiler
+isolation technique that can help de-couple the plugin host and the
+specifics of any given plugin.
+
+In the constantly evolving landscape defining which dependency versions
+are in use within which versions of common DCC tools, avoiding the need
+to matrix-compile and juggle plugin search paths per-host would be
+massively beneficial.
+
+## Options considered
+
+### Option 1
+
+`C++` and `Python` support only.
+
+**Estimated Cost**: Small
+
+#### Pros
+
+-   Simplifies development.
+
+#### Cons
+
+-   Leaves use within languages other than these a potentially
+    challenging exercise left to the reader.
+-   Compiler isolation support would need to be handled as an add-on
+    through an additional layer such as a `C` suite, `protobuf` or
+    similar.
+
+### Option 2
+
+`C` core library, bound to `C++` and `Python` as appropriate.
+
+**Estimated Cost**: Large
+
+#### Pros
+
+-   Makes future bindings or adoption in `C`-based languages far
+    simpler as all core API functionality is inherently available
+    through idiomatic `C`.
+-   Inherently presents a stable ABI isolation layer between hosts and
+    plugins.
+
+#### Cons
+
+-   The data model of OpenAssetIO _requires_ `map` and `optional` types.
+    Implementing these in pure `C` either requires significant coding,
+    or would add additional third-party dependencies. Either of these
+    add significantly to code complexity, maintenance overhead and
+    project risk.
```

Comment:
I disagree on the "significant coding" or "third-party depencies" point, and the subsequent risk. hash tables are well known solutions for maps, and it's not hard to design an `optional` surrogate. I'd word this point instead as 

The data model of OpenAssetIO _requires_ `map` and `optional` types. The lack of standardization in C around these concepts means a local implementation of these for C is required, and a mapping from this implementation to C++ and Python concepts introduces runtime and maintenance overhead.

Reply:
Reworded [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/8c053b05bd8dda23fb133640698a7900b2401514), also keeps the risk part, which is certainly a consideration from our p.o.v.